### ISSUE TITLE
Finding max_ply and reduce_ply

### ISSUE THREAD

**Issue by jjoshua2** [syzygy1/tb#13]
The biggest fishtest contributor has a machine with 1tb of ram and is attempting to build 7 piece TB.
https://groups.google.com/d/msg/fishcooking/GgsALC0GOxs/duzznHXHAQAJ

I think I know how to fix the permute indexing that crashed him so far [like so](https://github.com/jjoshua2/tb/blob/d7950859c668bcc269473b01ed86f15946527af6/src/permute.c). 

But I'm not sure about how you found max_ply of 509 and reduce_ply of 122 and reduce_ply_red in rtbgen.c. Is this related to what you said:
"something needs to be done as well to handle very large DTZ values. That only affects a few tables, though. (My idea is to simply divide them by 2 for DTZ > 50. Then everything still fits.)"

**Comment by jjoshua2** [comment]
I just realized atbgen.c and another one is for atomic and suicide which we're not interested in. But types.h has the ones for regular I think. These probably need to be increase?
`const int MAX_MOVES      = 256;
const int MAX_PLY        = 120;
const int MAX_PLY_PLUS_6 = MAX_PLY + 6;`

**Comment by jjoshua2** [comment]
I think the large dtz values is handled in rtbver.c. It supports about up to 256 but there are some special cases so its a bit less. I'm guessing if there's an overflow it uses `DTZ_ERROR 254`?

**Comment by jjoshua2** [comment]
In generics.c does `assume(n >= 3 && n <= 7); \
  long64 end = thread->end >> 6; \` have to also update to ` >>7`

**Comment by jjoshua2** [comment]
I'm guessing TBMAX_PIECE and TBMAX_PAWN will also have to be raised? It looks like if we hit them we'll get "limit too low!" error...

**Comment by noobpwnftw** [comment]
During the generation of KQQRBNvK, MAX_CANDS in permute.c overflowed, I don't know which value should I raise to.

**Comment by jjoshua2** [comment]
I'm wondering if idx << 6 should also be changed to idx << 7. Not sure about MAX_CANDS, I can't figure out where it came from unlike 5040 which is just 7 factorial.

**Comment by jjoshua2** [comment]
@noobpwnftw I think an easy to find MAX_CANDS would be after line 461 `int num_cands = est_data.num_cands;` put a print statement and see how big it ever gets?

**Comment by jjoshua2** [comment]
I was just posting that `mask_a1h1 = (mask_a1h1 << 6) | 0x07;` was scaled by numpieces and the others weren't so I think they are ok.

**Comment by jjoshua2** [comment]
@noobpwnftw Was it trylist or idx_cache that got index out of bounds?

**Comment by noobpwnftw** [comment]
https://github.com/syzygy1/tb/blob/2ba7088c8b483f41b79616ab1f4175da54c76007/src/permute.c#L520

This line, however I always suspect that this whole compression thing isn't simply going to work if we are just fiddling with the numbers.

**Comment by jjoshua2** [comment]
line 727 `uint32 dsize = num_segs * seg_size;` thats easy to see overflowing. why not just convert to uint64?

**Comment by jjoshua2** [comment]
long64 tb_size becomes size when passed to a method, which then sets `seg_size = size / num_segs;` so if tb_size is 64 bits then it makes sense that dsize has to be.

**Comment by noobpwnftw** [comment]
```
#define TBMAX_PIECE 650
#define TBMAX_PAWN 861
#define HSHMAX 12
```
I guess these are correct values for 3-7 men TBs.

**Comment by jjoshua2** [comment]
What is hshmax? Did you get those values from lomonostov 7 piece statistics? Slightlty trickty to convert from DTM to DTZ. The approximate increase looks about right. Once all are generated we should know the max dtz that we ever saw. I don't think there should be much penalty for setting these too high.

**Comment by noobpwnftw** [comment]
That's for how many bits needed to hold the tables. And the TBMAX_* numbers are nothing but how many total piece combinations possible with or without pawns.

https://github.com/syzygy1/tb/blob/0fa379c4406b3157fff15712e5ce1f712c23b5cf/src/permute.c#L552
It still crashed here.

**Comment by jjoshua2** [comment]
I just figured out MAX_CANDS is maximum number of candidates to try, which is why trylist is set to that (default 30). I think line  `long64 idx_cache[MAX_CANDS];` 478 could instead be `long64 idx_cache[num_cands];`
I assume it was idx_cache that was out of index?

**Comment by noobpwnftw** [comment]
Nope, I think it has something to do with encoding, all KQRBN+X case will fail, i.e. KQQRBNvK KQRRBNvK, etc. I will see how can I fix this when I have time.

**Comment by jjoshua2** [comment]
Maybe we can skip the 6v1 then? It's the least useful and 5v1 worked and 4v2, so 5v2 should also work, but I can see why 6v1 would be different.

**Comment by noobpwnftw** [comment]
As you can see, the side of this X is not relevant, so if it doesn't work with 6v1 then it won't work on 5v2, 4v3 either.

**Comment by syzygy1** [comment]
It seems I have accidentally "unwatched" this repository (probably a long time ago already).

MAX_CANDS = 30 = 6 * 5, so it has to be 7 * 6 = 42 now.
The TBMAX_PIECE and TBMAX_PAWN values are correct, and I also had HSHMAX = 12 in a local copy on my PC.

**Comment by syzygy1** [comment]
The current compression algorithm has a problem when the data has more than about 240 unique values. This was not a problem for 6-piece DTZ, since max DTZ is about 243 moves (I think), which is mapped to 1..100 ply (non-cursed) and 1..(243-50) moves (cursed). So it just needs 193 values. (It is fine that the ranges overlap, since probing WDL tells us which range is the correct one for a particular position. Btw, probing WDL also tells us the correct sign.)

For 7 pieces we have max DTZ = 517. This means a range from 1 to 467 for the cursed values. If we divide this by 2, it fits. But then probing cursed positions becomes a hassle, since we lose 1 bit of accuracy in the DTZ value.

So I would prefer modifying the compression algorithm to deal with 16-bit values. This is quite trivial to do (except that it involves a lot of code duplication, but so be it).

Unfortunately, this requires a lot of memory:
- the "white" half of the generated table needs to be loaded in RAM as 16-bit values: 462*64^5*2 = 924 GB (the memory allocated for generated white+black data with 1 byte per position can be reused for this);
- the permuted data must be stored at 2 bytes per position (max. 31332*61*60*59*58*2 = 731GB).

In total this gives 1655GB (plus a bit more for other stuff), which is a lot. Probably too much?

Fortunately, the permuted data is constructed sequentially (*), so it can just be written to disk instead of stored in RAM. When the permutation phase is finished, it can be loaded into RAM and the compression can be performed.

(*) Well, more or less. It is done in parallel, each thread sequentially working on its own range. This complicates matters a bit but is not a problem that cannot be solved.

**Comment by syzygy1** [comment]
The current generator uses LZ4 compression for writing intermediate data to disk. This compression is not very good but it is fast. The idea is that writing/reading LZ4-compressed data is faster than writing/reading uncompressed data. I haven't really measured if it helps and for 6-piece TBs it doesn't matter too much (I certainly spent more time on integrating the LZ4 compression code than it saved me when generating the TBs).

For 7-piece TBs the amount of data (and the number of tables) is much bigger, so it is probably worth it to look into this better. I guess it should pay off to use a slower but better compression algorithm and run it in parallel, each thread compressing a chunk of data and writing it to disk.

**Comment by noobpwnftw** [comment]
@syzygy1 
Losing 1 bit of accuracy in cursed DTZ value actually does virtually no harm, since engines don't really need to depend on it being precise and it only matters when there is a 3-fold.

But I still support to use 16-bit values for the sake of perfection, you can make a separate tool that only for producing 16-bit DTZs, so in such case the current generator can save the intermediate results and we use the tool to produce the final files. This way you don't have to make it look so duplicated:)

I think the current compression works just fine, parallel compression may be faster but the there is a cap on how fast results can be written on disk. It would require much more memory per thread and in fact there is few alternative that is even remotely close to the amazing throughput of LZ4.

**Comment by syzygy1** [comment]
Do we know how many 7-piece tables have max DTZ >= 290 moves? Probably not many and perhaps there are no such pawn tables at all.

The generator does not need to know the DTZ of subtables, so it would be possible to simply skip generating DTZ tables with DTZ >= 290 until I have extended the generator (which shouldn't take that long). If there are just a couple of files there is not much harm in having to regenerate them later to get the DTZ files.

That means the more urgent question is whether to switch the pawn generator to compressing by rank instead of per file in the hope of improving compression (which could be tested by generating some 6-piece TBs and comparing). I think I will add this possibility eventually, but it could take a while and I don't want to promise a date. So maybe you prefer to continue with the generation of 7-piece tables in their current format. The only "risk" is that the size of the resulting pawn tables is a few percent bigger than the size of rank-based tables. The good thing is that the current format will continue to be supported by engines, since it is the format of the 6-piece TBs that have been around since 2013.

I may eventually implement further improvements such as making use of same-piece symmetry (halving memory requirements for a table like KQQRvKBN and probably speeding up generation) and such changes could also result in slight changes in the resulting files (small things may influence the ordering of pieces etc.). We can probably accept that a later version of the generator produces slightly different tables.

**Comment by noobpwnftw** [comment]
I guess very few. I will add this under `if (generate_dtz) {` so that generator saves what it has and won't try to generate DTZ tables.
```
    if (!save_to_disk && (lcw_ply >= 290 || lcb_ply >= 290)) {
      store_table(table_w, 'w');
      store_table(table_b, 'b');
      exit(0);
    }
```
Compared to pawn generator improvements, I think it is more promising to have same-piece symmetry since a big portion of combinations in 7-piece would have the same pieces in one side.
Also we need to have pawnless tables before generating the pawnful ones, that makes this one more "urgent" in my sense.

Please take your time on everything, I'm not really into some kind of a race to get something done for the record or some credit, just worried because what used to limit what we can do is no longer there and we shouldn't fall behind.

**Comment by syzygy1** [comment]
I've increased the array sizes by 1, but I don't think this can be the problem.

**Comment by syzygy1** [comment]
This should fix it: https://github.com/syzygy1/tb/commit/c2a8c5d01aa2ed3e2f52c96d8d960277b38a5af0

Since pw[1] is only non-zero if there are 2 or more white pawns, this explains why tables with a single pawn are OK. (It may still be a good idea to test this, e.g. by just regenerating the table and compare the statistics without doing compressing the data (option -g).)

**Comment by syzygy1** [comment]
My apologies! I now see that you had already caught this problem in your PR.

**Comment by noobpwnftw** [comment]
@syzygy1 No problem, I used your code during generation, not my PR fork.
Now I will regenerate all 6v1 with 2 or more pawns and compare the output.

I have question about the other commit, were they related to the problem too or it was just the overflow alone?

**Comment by syzygy1** [comment]
The pfactor etc. being 32 bit instead of 64 bit is not a problem. However, it may anyway be better to make them 64 bit just to avoid having to think about it (and perhaps avoid problems in the future).

Yes, > 5 (or rather 5-1) should be impossible with 7 pieces. The most critical one is KQQQQQvK where they are accessed with 5-1=4.

**Comment by syzygy1** [comment]
Instead of regenerating all tables with a single pawn you could check that the cast to 64 bit does not change the outcome of that calculation. (Which seems to be clearly the case since the only problem is pw[1] << 30 and pw[1] is non zero only if piece "1" is a white pawn. For that we need at least two pawns (since the leading pawn is in position 0).)

**Comment by noobpwnftw** [comment]
I see, and KQQQQPvK has shown no difference after rebuilt.
I have removed those possible incorrect ones from FTP and have them regenerated.

**Comment by syzygy1** [comment]
Same for KQQQPPvK:
ftp://112.73.74.24/pub/syzygy/7men_testing/KQQQPPvK.txt
ftp://112.73.74.24/pub/syzygy/7men_bad_backup/KQQQPPvK.txt

**Comment by syzygy1** [comment]
I thought I had corrected all the shifts, but apparently not.
https://github.com/syzygy1/tb/commit/dd38aa65cd1497fbebca1c8f68a8c40b4034a375

This should again have affected only tables with two or more pawns (including PvP).

**Comment by noobpwnftw** [comment]
Affected files has been moved into folder 7men_bad_backup2, regeneration has started with new generator build.

**Comment by syzygy1** [comment]
The refactor branch now has code for dealing with DTZ >= 290. It only works for rtbgen so far. (rtbgenp does not even compile yet.)

**Comment by noobpwnftw** [comment]
reduce_tmpl.c is missing in the branch.
From Lomonosov results, I suspect KRBNKQN might be the only one that needs special treatment(Pawn movements reset DTZ).

**Comment by syzygy1** [comment]
Let's hope for a few more :-)

Missing file added.

**Comment by syzygy1** [comment]
This new rtbgen should behave like the old one. If max DTZ >= 290 it will use about the same amount of memory as normal (both with -d and without -d) but save more intermediate results to disk. The main visible difference is that it will try btm/dtz permutations before wtm/dtz permutations (if DTZ >= 290).

I have tested it by changing MAX_NARROW to 50 and generating a few TBs. The generated files were identical to the existing ones.

**Comment by noobpwnftw** [comment]
Now I'm testing the new rtbgen by building KRBNKQN, using the build prior to the commit 3e20b701558169c3a06b59bf2bc7715a76ff901e, it shouldn't matter in this particular case.

For some reason, the master branch generator failed to compress KRBBNvKR DTZ, I didn't keep intermediate stdout log but I will try to figure out what happened afterwards.

**Comment by syzygy1** [comment]
Thanks a lot, I will investigate. That the master branch crashes while compressing is quite surprising. The other problem is probably a stupid oversight.

**Comment by noobpwnftw** [comment]
Previous step(btm/wdl compression)
```

best order: 2
best permutation: 6 4 12 2 14 3 3
time taken =   9:20.574
compressing data for btm/wdl
calc_size: 313688938
bits = 6; blocks = 5349288 (0); size = 354175728
bits = 5; blocks = 9953886 (3); size = 340677570
real_num_blocks = 9930080; num_blocks = 9930083
idxbits = 19
num_indices = 374240
fixing cursed black losses.
```

Full log:
[log.txt](https://github.com/syzygy1/tb/files/1921495/log.txt)

**Comment by syzygy1** [comment]
Strange that this happens with a table that doesn't very look special.

Just to be sure: could you delete the .depsr directory, make clean, recompile, and check that the rtbgen binary is still the same?

I've ran a test with -fsanitize=undefined and only got some misaligned accesses (which are by design, although I should probably fix them at some point), signed integer overflows in myrandom() (which is harmless) and an overflow when initialising binomial[][] (in entries that aren't being used).

I've fixed the initialisation overflow, but that should not make a difference.

If it does not slow down too much, you could perhaps try building KRBBNvKR with rtbgen compiled with -fsanitize. (Comment out the second CFLAGS line and uncomment the third in the Makefile.)

**Comment by syzygy1** [comment]
I may have fixed the KRBNvKQN problem.

The compression code still contained a check to guard against more than 255 symbols to be compressed and exited with "error" if this value was exceeded.

This does not explain why the generator got stuck, but at the moment I don't really see how this check could not have been triggered exactly at the point where the generator stopped for you, so there does seem to be a good chance that this is the problem.

**Comment by syzygy1** [comment]
http://centaur.reading.ac.uk/4608/1/2006c_ICGA_J_H_Chess_Endgame_News_29.2.pdf

The 16-bit compressor seems to be needed for at least 4 tables:

KRRNvKRR -> 290 moves
KQBNKQN -> 317 moves
KQBNKQB -> 330 moves
KRBNKQN -> 517 moves

**Comment by noobpwnftw** [comment]
It seems that the ubsan version is way too slow, I will try to look for other methods to narrow down the problem. (Skipping WDL, doubling size of each mallocs in compress.c) and see if they make any difference.

**Comment by noobpwnftw** [comment]
Yes, it still happen without compressing WDL, and I may have "fixed" it by some fiddling https://github.com/syzygy1/tb/blob/2b8c8649646344a428e1c6c832c73213f6b9ce66/src/compress.c#L30-L31 raising 32 to 64 solved the crash, but I do not know what's behind these values.

**Comment by syzygy1** [comment]
Aha! It may be related to the Huffman code having  codewords with lengths of over 32 bits. I did think of this possibility yesterday, but I did not see how it could cause a crash while generating the table. But these arrays might explain.

I probably have to limit the length of codewords to 32 bits, since the decompression code relies on that (I think).

The reason that this never went wrong before is probably that with 6-piece tables the rarest values are much less rare (relatively speaking) than with 7-piece tables.

I will try to come up with a way to check that the tables generated so far don't have the same problem. (But there shouldn't be any problems with WDL tables because of this, so any problems there may exist will be isolated.)

Thanks for the excellent debugging work!

**Comment by noobpwnftw** [comment]
These arrays are located in a memory allocated by malloc, like the following:
[header1][memory1][header2][memory2][header3][memory3]....
If there is an overflow somewhere, it will overwrite next header and cause the next malloc to fail due to internal sanity checks, I don't think any other tables are likely affected because it would result in the same crash. It may even have some alignments in between that small overflows can be harmless.

**Comment by syzygy1** [comment]
If the number of "empty" bits is 32 or higher, 4 more bytes of compressed data are read. So the decompression code will have problems with codewords longer than 32 bits. That could be fixed, but it seems better to just limit codeword length. I will do that later today.

**Comment by syzygy1** [comment]
So I have to implement the package-merge algorithm.
http://www.diku.dk/~jyrki/Paper/ISAAC95.pdf

If there turn out to be many tables among those already generated with too long huffman codes, I think it should be possible to recompress. Of course we then first have to be sure that they have not been corrupted.

**Comment by noobpwnftw** [comment]
I have re-tried the refactor branch, it still got stuck, I think it is in convert_est_data_piece threaded code. All threads seems alive but waiting on the barriers.

**Comment by syzygy1** [comment]
Yes, it will be easy to check which files are affected by checking max_len. Unfortunately I haven't been able yet to implement a length-limited huffman code builder.

**Comment by noobpwnftw** [comment]
The refactor branch still hang after my last attempt of change.

From GDB I can see all threads either waiting on barrier_start or on barrier_end of convert_est_data_piece work right after 

> compressing data for btm/wdl
> calc_size: 14375357332
> bits = 6; blocks = 226819177 (0); size = 15041919612
> real_num_blocks = 226411747; num_blocks = 226411747
> idxbits = 15
> num_indices = 11975655
> Verifying reconstructed table_b based on collected statistics.
> find optimal permutation for btm/dtz

Looks like some overflow on the stack.

**Comment by jjoshua2** [comment]
Small request. since you have so many cpu cores maybe you could add
calculating sha1 of all the tb and host the sha1 file? Then we can verify
correct download of so much data.

**Comment by syzygy1** [comment]
tbcheck can be used to check the internal checksum of each file.

**Comment by noobpwnftw** [comment]
@syzygy1 
I will give another try of refactor branch with more printfs added and only -z option.
It might be easier if I give you access to that machine? I've sent an email to you.

**Comment by noobpwnftw** [comment]
From debugging I have confirmed that it haven't reached the compression part, or it should have done printf of at least one permutation before hanging in the next loop of convert_est_data_piece.

**Comment by syzygy1** [comment]
So the call to run_threaded() in estimate_compression_piece_u16() results in all threads being blocked in line 234 for threads.c? (pthread_barrier_wait(&barrier_end))

**Comment by syzygy1** [comment]
The point of pthread_barrier is that the threads wait until all threads get there (numthreads in total). If all threads did get there, the barrier isn't doing what it was supposed to be doing. So then it probably got corrupted somehow.

**Comment by noobpwnftw** [comment]
These are known good things:
1. WDL compressed fine, no memory corruption or it would fail like the other one in some following malloc.
2. Reconstruction should be fine, verification found no errors.
3. Point in time debugging shows all threads alive, waiting at worker-> convert_est_data_piece, thus not some exit(1) called, even before removing that num_vals < 256 check.
4. I have checked the code with alloc_work and fill_work in permutation code but I can't see anything wrong.

Known bad things:
1. I can see some threads waiting on barrier_start, some on barrier_end.
2. I tried to debug the pthread mutex, it seems corrupted(structure shows some random value).
3. Building with -z failed because that extra free called in my PR.

**Comment by syzygy1** [comment]
Do you know if the 16-bit DTZ permutation also hangs for other 7-piece tables with lower DTZ? (So after lowering MAX_NARROW?)

**Comment by noobpwnftw** [comment]
I have not yet tried that on tables with 7-piece, or involving a reconstruction.

**Comment by syzygy1** [comment]
If it hangs without reconstruction, we know it's not the reconstruction that causes it.

**Comment by syzygy1** [comment]
I noticed a problem in permute.c where copybuf is not checked for being non-zero. So it will crash with -z if there is no reconstruction.

**Comment by syzygy1** [comment]
The copybuf problem is fixed now.

**Comment by noobpwnftw** [comment]
I see it has threaded work within verify_stats which means the barriers worked fine at that moment and gets corrupted afterwards, thus I wonder what might caused it. Doesn't seem to be the reconstruction part.

**Comment by syzygy1** [comment]
Interesting :)

**Comment by syzygy1** [comment]
Hopefully fixed now.
The >32 bit Huffman codewords still need to be addressed.

**Comment by noobpwnftw** [comment]
I see, thanks.
BTW, I've sent you an email about accessing to my machine, if that helps you with development.

**Comment by syzygy1** [comment]
It may help with debugging, but hopefully it is not needed anymore.
I'll try to implement length-limited Huffman codes over the weekend.

**Comment by syzygy1** [comment]
Huffman codewords are now limited to 32 bits.

The old code is used first. If it results in codewords that are >32 bits long, a new length-limited code is constructed. This is to avoid unnecessary changes in the TB files produced.

I have tested the new code by commenting out the first two lines in create_code() (which calls the old code first) and verifying the resulting TB files (with rtbver).

KRBBNvKR should now build correctly. Unless the problem is a different one :-)

I will now look into a way of checking max_len for existing files (also for 6-piece TBs...).

**Comment by syzygy1** [comment]
Check max_len with ./rtbver -h KQRvKBN (and a perl script).
It will print the max_len values for WDL and DTZ and add WARNING!!! if one is >= 32.

Although 32 bits should be handled correctly by the decompression code, it overflows an array in the generator, so the file could potentially be corrupted. (I have increased the array size now to allow 32 bits.)

**Comment by syzygy1** [comment]
If ./rtbver prints a 0 for max_len, it is because the table compressed to a constant value (e.g. the wtm half of KQQQQQvK).

KQNNNvKQ.rtbz and KQRRBvKQ.rtbz have max_len of 29, so there are probably some tables with 32 or higher.

**Comment by syzygy1** [comment]
All 3-6 piece pawnless tables are OK. Maximum max_len is 28 for KQRvKQN.rtbz.

**Comment by syzygy1** [comment]
Interestingly, regenerating KQRvKQN with a limit of 25 results in a TB file of exactly the same size (but indeed max_len = 25). The reported "calc_size" is exactly 1 byte higher (the compressed size before splitting into blocks).

**Comment by syzygy1** [comment]
I have now done the same for rtbverp and added an option to run.pl (--huffman instead of --generate).
The pawnful max_len values seem to be generally lower than the pawnless ones, so hopefully KQRRBvKR is the only affected table.

**Comment by noobpwnftw** [comment]
I have checked all 6v1 ones, no tables affected.

**Comment by noobpwnftw** [comment]
Actually I'm at the last 6v1 pawnful table KPPPPPvK, it has been building single-threaded for a long while because of https://github.com/syzygy1/tb/blob/2b8c8649646344a428e1c6c832c73213f6b9ce66/src/tbgenp.c#L782-L783

Is this intended or it may be possible to parallelize a bit?

**Comment by syzygy1** [comment]
It is intended because the slices are so small here. You could try changing the 3 to 2 in the slice_threading_high line, but it might only slow down.

The solution would be to let different threads work on different slices, but I have not implemented that because it is tricky and it matters for only a few tables.

**Comment by noobpwnftw** [comment]
Ok, I will just wait it out, there are only few of those anyways.

About the codeword limit you lowered but get the same compressed size, I think that proves your Huffman tree is optimal, that 3 bits probably got into another symbol of equal weight. If compressed size can be further lowered by splitting codewords for reuse elsewhere, then it means the original tree wasn't optimal.

**Comment by syzygy1** [comment]
By the time you get to KPPPPvKP and KPPPvKPP there may be an improvement that speeds up the generation of files b, c, d.

**Comment by syzygy1** [comment]
To do KPPPPPvK in parallel, I could split the range 0..pawnsize in calc_pawn_table() into pieces corresponding to assignments of the pawns to files. Slices can be dependent only if each pawn is in the same file in both slices.

The first pawn is already fixed in one of files a, b, c, d. The other 4 pawns are in 8 possible files, so that gives 8^4 = 4096 work units which can be processed in parallel. That is plenty. It could even be done for tables with 4 or 3 pawns in total (512 and 64 work units), but then table reductions complicate things a bit more (I could look into that later).

If the current KPPPPPvK generation is going to be a matter of many days or longer, you might want to skip it for the moment and continue with 5v2.

**Comment by syzygy1** [comment]
On the other hand, if it is making progress, maybe just let it run for the moment. I'm not sure yet how easy it will be to run some parts of the generator in parallel that were not intended to run in parallel so far.

**Comment by noobpwnftw** [comment]
I have two machines, so I will just keep it running(now building a5), it will probably need a week or so while the 5v2 building can proceed in parallel without problem.

**Comment by syzygy1** [comment]
Compression of pawnful tables crashes in the refactor branch at the moment.
When I have fixed that, I am very close to getting KPPP..vK to run multithreaded in a special subbranch.
It will have to wait till tomorrow.

**Comment by noobpwnftw** [comment]
I just checked, it was using the master branch, 😄 . And I have built KRBNvKQN successfully, uploading.

**Comment by syzygy1** [comment]
Yes, I assumed you were still using the master branch for rtbgenp. The refactor rtbgenp did not compile until yesterday. Now it does, but there is still something wrong somewhere. (I have now fixed one problem but there is another one.)

Good to hear that KRBNvKQN worked!

**Comment by syzygy1** [comment]
I think rtbgenp from the refactor branch should work now.

You can also try tbgenp from the ppppp branch. It should be much faster for KPPPPPvK etc. It will exit if reductions are needed (which shouldn't be the case for KPPPPPvK, KPPPPvKP and KPPPvKPP).

**Comment by syzygy1** [comment]
It seems the ppppp branch is not yet entirely bug free. The statistics for KPPvKPP differ.

Even more worrying is that the "old" KPPvKPP statistics are not symmetrical:
########## KPPvKPP ##########

White to move:

3602411932 positions win in 1 ply.
897476 positions win in 2 ply.
161637352 positions win in 3 ply.
73956 positions win in 4 ply.

...

Black to move:

3602411920 positions win in 1 ply.
897476 positions win in 2 ply.
161637364 positions win in 3 ply.
73956 positions win in 4 ply.


On the other hand, the "old" overall statistics are identical for wtm and btm:
3799478076 positions are wins.
148184 positions are cursed wins.
1166785852 positions are draws.
152224 positions are cursed losses.
2041799060 positions are losses.

3799478076 positions are wins.
148184 positions are cursed wins.
1166785852 positions are draws.
152224 positions are cursed losses.
2041799060 positions are losses.

But this is not the case for the "new" overall statistics:
3799477599 positions are wins.
148184 positions are cursed wins.
1166786502 positions are draws.
152224 positions are cursed losses.
2041798887 positions are losses.

3799477938 positions are wins.
148184 positions are cursed wins.
1166786769 positions are draws.
152224 positions are cursed losses.
2041798281 positions are losses.


My dtm generator produced this (for both sides):
3799625764 positions are wins.
1166786492 positions are draws.
2041951140 positions are losses.

So different numbers of wins/draws/losses...

**Comment by syzygy1** [comment]
Nalimov again gives different numbers, but they don't seem to be comparable (my numbers have to be divided by 4, but then the total number of positions still doesn't seem to match).
nalimov wins: 3904409432
nalimov draws: 1188051112
nalimov losses: 2065089316
nalimov total: 7157549860

My KQPvKQP, KRPvKRP, KBPvKBP and KNPvKNP statistics look normal.

**Comment by syzygy1** [comment]
The ppppp branch also shows a small deviation on KPPPPvK: 2 btm loss-in-4 positions less.
The rest is the same, so this is strange: where did those 2 positions go?

Black to move:


0 positions are wins.
5920344 positions are draws.
7410806854 positions are losses.

340344 positions lose in 0 ply.
11917728 positions lose in 1 ply.
7396110960 positions lose in 2 ply.
674352 positions lose in 3 ply.
1682398 positions lose in 4 ply.

The "old" numbers:
Black to move:


0 positions are wins.
5920344 positions are draws.
7410806856 positions are losses.

340344 positions lose in 0 ply.
11917728 positions lose in 1 ply.
7396110960 positions lose in 2 ply.
674352 positions lose in 3 ply.
1682400 positions lose in 4 ply.

**Comment by syzygy1** [comment]
The problem in the ppppp branch should be solved now. There was still a global variable that should have been turned into a thread-specific variable.
KPPPPPvK should be regenerated.

The other problem I have not yet looked at in depth.

**Comment by syzygy1** [comment]
Hmmm, the fixed ppppp generator does produce symmetrical statistics on KPPvKPP:

White to move:

3602411920 positions win in 1 ply.
897476 positions win in 2 ply.
161637364 positions win in 3 ply.
73956 positions win in 4 ply.
22385016 positions win in 5 ply.

3799478076 positions are wins.
148184 positions are cursed wins.
1166785852 positions are draws.
152224 positions are cursed losses.
2041799060 positions are losses.

Black to move:

3602411920 positions win in 1 ply.
897476 positions win in 2 ply.
161637364 positions win in 3 ply.
73956 positions win in 4 ply.
22385016 positions win in 5 ply.

3799478076 positions are wins.
148184 positions are cursed wins.
1166785852 positions are draws.
152224 positions are cursed losses.
2041799060 positions are losses.

The files produced are identical to my current files, so it seems my KPPvKPP.txt was simply out of date (but strangely they were generated on the same day).

The statistics are still not identical to the W/D/L number produced by my dtm generator, so not all is fine yet. But the problem may well be in my DTM generator.

**Comment by Saugstrahler** [comment]
Talking about DTM generator....is it available now for the 6-men or are you still fiddling around with the final format?

**Comment by noobpwnftw** [comment]
I have rebuilt the KPPPPPvK table.

**Comment by syzygy1** [comment]
I have found and fixed a problem in my DTM generator that was probably responsible for my incorrect KPPvKPP DTM statistics. At least, I found a few more tables with problems and of those I have now succesfully regenerated KBPPvKP.

The problem does not affect the WDL/DTZ generator.

**Comment by syzygy1** [comment]
I've merged refactor into master and opened a new issue: https://github.com/syzygy1/tb/issues/25
